chainstate/src/detail/best_chain_candidates.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeSet;
+
+use thiserror::Error;
+
+use super::chainstateref::ChainstateRef;
+use crate::TransactionVerificationStrategy;
+use chainstate_storage::BlockchainStorageRead;
+use chainstate_types::{BlockIndex, GenBlockIndex, PropertyQueryError};
+use common::{chain::Block, primitives::Id, Uint256};
+use utils::tap_error_log::LogError;
+
+#[derive(Eq, Copy, Clone, Debug)]
+pub struct BestChainCandidatesItem {
+    pub chain_trust: Uint256,
+    pub block_id: Id<Block>,
+}
+
+impl BestChainCandidatesItem {
+    fn from_block_index(block_index: &BlockIndex) -> Self {
+        BestChainCandidatesItem {
+            block_id: *block_index.block_id(),
+            chain_trust: block_index.chain_trust(),
+        }
+    }
+}
+
+impl Ord for BestChainCandidatesItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        (&self.chain_trust, &self.block_id).cmp(&(&other.chain_trust, &other.block_id))
+    }
+}
+
+impl PartialOrd for BestChainCandidatesItem {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for BestChainCandidatesItem {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == std::cmp::Ordering::Equal
+    }
+}
+
+/// A collection of candidates for the best chain.
+#[derive(Debug)]
+pub struct BestChainCandidates(BTreeSet<BestChainCandidatesItem>);
+
+impl BestChainCandidates {
+    /// Collect candidates for the best chain that have chain trust bigger than or equal to
+    /// the specified one.
+    /// Only consider branches that start above the minimum height where reorgs are allowed.
+    pub fn new<S, V>(
+        chainstate_ref: &ChainstateRef<S, V>,
+        min_chain_trust: Uint256,
+    ) -> Result<BestChainCandidates, BestChainCandidatesError>
+    where
+        S: BlockchainStorageRead,
+        V: TransactionVerificationStrategy,
+    {
+        let best_block_index = chainstate_ref.get_best_block_index().log_err()?;
+        let min_height_with_allowed_reorg = chainstate_ref.get_min_height_with_allowed_reorg()?;
+
+        let block_ids_by_height = chainstate_ref
```

Comment:
Please add a comment explaining (again at the caller) that this is linear.

Reply:
I've added the comment.

Also, I've just realized that the call to `last_common_ancestor` below is more expensive than I thought - it iterates towards the common ancestor block by block instead of taking bigger jumps via `get_ancestor`. So, first of all, I replaced the call to `last_common_ancestor` with `last_common_ancestor_in_main_chain`. Currently, the two functions basically do the same. But the latter can actually be optimized so that it takes bigger jumps via `get_ancestor`, so I wrote a TODO about it.